crates/sui-types/src/gas_model/gas_v2.rs

```diff
@@ -0,0 +1,501 @@
+
+    fn move_gas_status(&mut self) -> &mut GasStatus<'a> {
+        &mut self.gas_status
+    }
+
+    fn bucketize_computation(&mut self) -> Result<(), ExecutionError> {
+        let gas_used = self.gas_status.gas_used_pre_gas_price();
+        let bucket_cost = get_bucket_cost(&COMPUTATION_BUCKETS, gas_used);
+        // charge extra on top of `computation_cost` to make the total computation
+        // cost a bucket value
+        // println!(
+        //     "GAS - Bucketize: gas_used {}, bucket cost {}",
+        //     gas_used, bucket_cost
+        // );
+        let gas_used = bucket_cost * self.gas_price;
+        if self.gas_budget <= gas_used {
+            self.computation_cost = self.gas_budget;
+            Err(ExecutionErrorKind::InsufficientGas.into())
+        } else {
+            self.computation_cost = gas_used;
+            Ok(())
+        }
+    }
+
+    /// Returns the final (computation cost, storage cost, storage rebate) of the gas meter.
+    /// We use initial budget, combined with remaining gas and storage cost to derive
+    /// computation cost.
+    fn summary(&self) -> GasCostSummary {
+        // compute storage rebate, both rebate and non refundable fee
+        let sender_rebate = sender_rebate(self.storage_rebate, self.rebate_rate);
+        assert!(sender_rebate <= self.storage_rebate);
+        let non_refundable_storage_fee = self.storage_rebate - sender_rebate;
+        // let summary = GasCostSummary {
+        GasCostSummary {
+            computation_cost: self.computation_cost,
+            storage_cost: self.storage_cost,
+            storage_rebate: sender_rebate,
+            non_refundable_storage_fee,
+        }
+        // };
+        // println!("GAS - gas_summary: {}", summary);
+        // summary
+    }
+
+    fn gas_budget(&self) -> u64 {
+        self.gas_budget
+    }
+
+    fn storage_gas_units(&self) -> u64 {
+        self.storage_cost
+    }
+
+    fn storage_rebate(&self) -> u64 {
+        self.storage_rebate
+    }
+
+    fn unmetered_storage_rebate(&self) -> u64 {
+        self.unmetered_storage_rebate
+    }
+
+    fn gas_used(&self) -> u64 {
+        self.gas_status.gas_used_pre_gas_price()
+    }
+
+    fn reset_storage_cost_and_rebate(&mut self) {
+        self.storage_cost = 0;
+        self.storage_rebate = 0;
+    }
+
+    fn charge_storage_read(&mut self, size: usize) -> Result<(), ExecutionError> {
+        // println!("GAS - charge_storage_read: size {}", size);
+        self.gas_status
+            .charge_bytes(size, self.cost_table.object_read_per_byte_cost)
+            .map_err(|e| {
+                debug_assert_eq!(e.major_status(), StatusCode::OUT_OF_GAS);
+                ExecutionErrorKind::InsufficientGas.into()
+            })
+    }
+
+    fn charge_storage_mutation(
+        &mut self,
+        new_size: usize,
+        storage_rebate: u64,
+    ) -> Result<u64, ExecutionError> {
+        // TODO: this is exclusively called from testing as it breaks too much
```

Comment:
What does this TODO mean?

Reply:
a bunch of tests are calling that API so for now I am leaving that around even though it is never called at runtime.
I'll fix it once we are all off this collective effort to fix test